src/main/java/com/googleintern/wfm/ruleengine/action/RuleValidation.java

```diff
@@ -0,0 +1,287 @@
+      new String[] {
+        "User Id",
+        "Workforce Id",
+        "Workgroup Id",
+        "Role Id",
+        "Skill Id",
+        "Expected Pool Assignments",
+        "Actual Pool Assignments"
+      };
+  private static final String PERCENTAGE_SIGN = "%";
+  private static final String CASE_POOL_ID_PREFIX =
+      Separator.CURLY_BRACKET_LEFT.symbol + "\"case_pool_id\":\"";
+  private static final String PERMISSION_SET_ID_PREFIX = "\",\"permission_set_id\":\"";
+
+  public RuleValidation(
+      Set<RuleModel> rules, List<UserPoolAssignmentModel> expectedUserPoolAssignments) {
+    this.rules = ImmutableSet.copyOf(rules);
+    this.expectedUserPoolAssignments = ImmutableList.copyOf(expectedUserPoolAssignments);
+    this.actualAssignedPermissionsByUserPoolAssignment =
+        assignPermissionsBasedOnGeneratedRules(expectedUserPoolAssignments);
+  }
+
+  public static void writeRuleValidationResultsIntoCsvFile(String outputCsvFilePath)
+      throws IOException {
+    File outputFile = new File(outputCsvFilePath);
+    Files.deleteIfExists(outputFile.toPath());
+    outputFile.createNewFile();
+
+    FileWriter outputFileWriter = new FileWriter(outputFile);
+    CSVWriter csvWriter = new CSVWriter(outputFileWriter);
+
+    ImmutableList<String[]> validationResult = convertRuleValidationResultsToListOfStringArrays();
+
+    csvWriter.writeAll(validationResult);
+    csvWriter.close();
+  }
+
+  public static double calculateRulesCoverage(
+      ImmutableSet<UserPoolAssignmentModel> usersWithWrongAssignedPermissions) {
+    double numberOfWrongAssignment = usersWithWrongAssignedPermissions.size();
+    double totalNumberOfUsers = expectedUserPoolAssignments.size();
+    return (totalNumberOfUsers - numberOfWrongAssignment) / totalNumberOfUsers;
+  }
+
+  public static ImmutableSet<UserPoolAssignmentModel> findUncoveredUsers() {
+    return expectedUserPoolAssignments.stream()
+        .filter(
+            user ->
+                ((!actualAssignedPermissionsByUserPoolAssignment.containsKey(user)
+                        && !user.poolAssignments().isEmpty())
+                    || (!user.poolAssignments()
+                        .equals(actualAssignedPermissionsByUserPoolAssignment.get(user)))))
+        .collect(ImmutableSet.toImmutableSet());
+  }
+
+  public static ImmutableSet<PoolAssignmentModel> findUncoveredPoolAssignments() {
+    ImmutableSet<PoolAssignmentModel> expectedAllPoolAssignments = findExpectedAllPoolAssignments();
+    ImmutableSet<PoolAssignmentModel> actualAllPoolAssignments = findActualAllPoolAssignments();
+    return ImmutableSet.copyOf(
+        Sets.difference(expectedAllPoolAssignments, actualAllPoolAssignments));
+  }
+
+  private static ImmutableSet<PoolAssignmentModel> findExpectedAllPoolAssignments() {
+    return expectedUserPoolAssignments.stream()
+        .map(user -> user.poolAssignments())
+        .reduce(ImmutableSet.of(), (union, current) -> Sets.union(union, current).immutableCopy());
+  }
```

Comment:
expectedUserPoolAssignments should be a parameter that is passing into this method instead of a private property of the class. 

Same below.